pallets/creator-staking/src/functions.rs

```diff
@@ -0,0 +1,309 @@
+        staking_info.total = staking_info.total.saturating_add(value);
+
+        Ok(())
+    }
+
+    /// An utility method used to unstake specified amount from an arbitrary creator.
+    ///
+    /// The amount unstaked can be different in case staked amount would fall bellow
+    /// `MinimumStakingAmount`. In that case, entire staked amount will be unstaked.
+    ///
+    /// `StakerInfo` and `CreatorStakeInfo` are provided and all checks are made to ensure that
+    /// it's possible to complete unstake operation.
+    ///
+    /// # Arguments
+    ///
+    /// * `staker_info` - info about staker's stakes on the creator up to current moment
+    /// * `staking_info` - general info about creator stakes up to current moment
+    /// * `value` - value which should be unstaked
+    /// * `current_era` - current creators-staking era
+    ///
+    /// # Returns
+    ///
+    /// If unstake operation was successful, given structs are properly modified and total
+    /// unstaked value is returned. If not, an error is returned and structs are left in
+    /// an undefined state.
+    pub(super) fn unstake_from_creator(
+        staker_info: &mut StakerInfo<BalanceOf<T>>,
+        stake_info: &mut CreatorStakeInfo<BalanceOf<T>>,
+        value: BalanceOf<T>,
+        current_era: EraIndex,
+    ) -> Result<BalanceOf<T>, Error<T>> {
+        let staked_value = staker_info.latest_staked_value();
+        ensure!(staked_value > Zero::zero(), Error::<T>::NotStakedCreator);
+
+        // Calculate the value which will be unstaked.
+        let remaining = staked_value.saturating_sub(value);
+        let value_to_unstake = if remaining < T::MinimumStakingAmount::get() {
+            stake_info.number_of_stakers = stake_info.number_of_stakers.saturating_sub(1);
+            staked_value
+        } else {
+            value
+        };
+        stake_info.total = stake_info.total.saturating_sub(value_to_unstake);
+
+        // Sanity check
+        ensure!(value_to_unstake > Zero::zero(), Error::<T>::CannotUnstakeZero);
+
+        staker_info
+            .unstake(current_era, value_to_unstake)
+            .map_err(|_| Error::<T>::UnexpectedStakeInfoEra)?;
+        ensure!(
+                // One spot should remain for compounding reward claim call
+                staker_info.len() < T::MaxEraStakeValues::get(),
+                Error::<T>::TooManyEraStakeValues
+            );
+
+        Ok(value_to_unstake)
+    }
+
+    /// Update the ledger for a staker. This will also update the stash lock.
+    /// This lock will lock the entire funds except paying for further transactions.
+    pub(super) fn update_ledger(staker: &T::AccountId, ledger: AccountLedger<BalanceOf<T>>) {
+        if ledger.is_empty() {
+            Ledger::<T>::remove(staker);
+            T::Currency::remove_lock(STAKING_ID, staker);
+        } else {
+            T::Currency::set_lock(STAKING_ID, staker, ledger.locked, WithdrawReasons::all());
```

Comment:
This will trigger to update lock on the balance even if `ledger.locked` was not changed. Looks not omptimal :(

Reply:
This happens only when we do `unstake`, where the actual lock doesn't change. In all the other places, where this function is used, we change locked amount.